File: clients/client-python/gravitino/api/rel/table_catalog.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.


from abc import ABC, abstractmethod
from contextlib import suppress
from typing import Optional

from gravitino.api.rel.column import Column
from gravitino.api.rel.expressions.distributions.distribution import Distribution
from gravitino.api.rel.expressions.sorts.sort_order import SortOrder
from gravitino.api.rel.expressions.transforms.transform import Transform
from gravitino.api.rel.indexes.index import Index
from gravitino.api.rel.table import Table
from gravitino.exceptions.base import (
    NoSuchTableException,
    UnsupportedOperationException,
)
from gravitino.name_identifier import NameIdentifier
from gravitino.namespace import Namespace


class TableCatalog(ABC):
    """The `TableCatalog` interface defines the public API for managing tables in a schema.

    If the catalog implementation supports tables, it must implement this interface.
    """

    @abstractmethod
    def list_tables(self, namespace: Namespace) -> list[NameIdentifier]:
        """List the tables in a namespace from the catalog.

        Args:
            namespace (Namespace): A namespace.

        Returns:
            list[NameIdentifier]: An array of table identifiers in the namespace.

        Raises:
            NoSuchSchemaException: If the schema does not exist.
        """

    @abstractmethod
    def load_table(self, identifier: NameIdentifier) -> Table:
        """Load table metadata by `NameIdentifier` from the catalog.

        Args:
            identifier (NameIdentifier): A table identifier.

        Returns:
            Table: The table metadata.

        Raises:
            NoSuchTableException: If the table does not exist.
        """

    def table_exists(self, identifier: NameIdentifier) -> bool:
        """Check if a table exists using an `NameIdentifier` from the catalog.

        Args:
            identifier (NameIdentifier): A table identifier.

        Returns:
            bool: `True` If the table exists, `False` otherwise.
        """
        with suppress(NoSuchTableException):
            self.load_table(identifier)
            return True
        return False

    @abstractmethod
    def create_table(
        self,
        identifier: NameIdentifier,
        columns: list[Column],
        comment: Optional[str] = None,
        properties: Optional[dict[str, str]] = None,
        partitioning: Optional[list[Transform]] = None,
        distribution: Optional[Distribution] = None,
        sort_orders: Optional[list[SortOrder]] = None,
        indexes: Optional[list[Index]] = None,
    ) -> Table:
        """Create a table in the catalog.

        Args:
            identifier (NameIdentifier):
                A table identifier.
            columns (list[Column]):
                The columns of the new table.
            comment (str, optional):
                The table comment. Defaults to `None`.
            properties (dict[str, str], optional):
                The table properties. Defaults to `None`.
            partitioning (Optional[list[Transform]], optional):
                The table partitioning. Defaults to None.
            distribution (Optional[Distribution], optional):
                The distribution of the table. Defaults to `None`.
            sort_orders (Optional[list[SortOrder]], optional):
                The sort orders of the table. Defaults to `None`.
            indexes (Optional[list[Index]], optional):
                The table indexes. Defaults to `None`.

        Raises:
            NoSuchSchemaException:
                If the schema does not exist.
            TableAlreadyExistsException:
                If the table already exists.

        Returns:
            Table:
                The created table metadata.
        """

    @abstractmethod
    def alter_table(self, identifier: NameIdentifier, *changes) -> Table:
        """Alter a table in the catalog.

        Args:
            identifier (NameIdentifier): A table identifier.
            *changes:
                Table changes (defined in class `TableChange`) to apply to the table.

        Returns:
            Table: The updated table metadata.

        Raises:
            NoSuchTableException:
                If the table does not exist.
            IllegalArgumentException:
                If the change is rejected by the implementation.
        """

    @abstractmethod
    def drop_table(self, identifier: NameIdentifier) -> bool:
        """Drop a table from the catalog.

        Removes both the metadata and the directory associated with the table from the
        file system if the table is not an external table. In case of an external table,
        only the associated metadata is removed.

        Args:
            identifier (NameIdentifier): A table identifier.

        Returns:
            bool: `True` if the table is dropped, `False` if the table does not exist.
        """

    def purge_table(self, identifier: NameIdentifier) -> bool:
        """Drop a table from the catalog and completely remove its data.

        Removes both the metadata and the directory associated with the table completely
        and skipping trash. If the table is an external table or the catalogs don't support
        purge table, `UnsupportedOperationException` is thrown. If the catalog supports to
        purge a table, this method should be overridden. The default implementation throws
        an `UnsupportedOperationException`.

        Args:
            identifier (NameIdentifier): A table identifier.

        Raises:
            UnsupportedOperationException: If the catalog does not support to purge a table.

        Returns:
            bool: `True` if the table is purged, `False` if the table does not exist.
        """
        raise UnsupportedOperationException("purgeTable not supported.")

File: api/src/main/java/org/apache/gravitino/rel/TableCatalog.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Referred from Apache's connector/catalog implementation
// sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java

package org.apache.gravitino.rel;

import java.util.Map;
import org.apache.gravitino.NameIdentifier;
import org.apache.gravitino.Namespace;
import org.apache.gravitino.annotation.Evolving;
import org.apache.gravitino.exceptions.NoSuchSchemaException;
import org.apache.gravitino.exceptions.NoSuchTableException;
import org.apache.gravitino.exceptions.TableAlreadyExistsException;
import org.apache.gravitino.rel.expressions.distributions.Distribution;
import org.apache.gravitino.rel.expressions.distributions.Distributions;
import org.apache.gravitino.rel.expressions.sorts.SortOrder;
import org.apache.gravitino.rel.expressions.transforms.Transform;
import org.apache.gravitino.rel.expressions.transforms.Transforms;
import org.apache.gravitino.rel.indexes.Index;
import org.apache.gravitino.rel.indexes.Indexes;

/**
 * The TableCatalog interface defines the public API for managing tables in a schema. If the catalog
 * implementation supports tables, it must implement this interface.
 */
@Evolving
public interface TableCatalog {

  /**
   * List the tables in a namespace from the catalog.
   *
   * @param namespace A namespace.
   * @return An array of table identifiers in the namespace.
   * @throws NoSuchSchemaException If the schema does not exist.
   */
  NameIdentifier[] listTables(Namespace namespace) throws NoSuchSchemaException;

  /**
   * Load table metadata by {@link NameIdentifier} from the catalog.
   *
   * @param ident A table identifier.
   * @return The table metadata.
   * @throws NoSuchTableException If the table does not exist.
   */
  Table loadTable(NameIdentifier ident) throws NoSuchTableException;

  /**
   * Check if a table exists using an {@link NameIdentifier} from the catalog.
   *
   * @param ident A table identifier.
   * @return true If the table exists, false otherwise.
   */
  default boolean tableExists(NameIdentifier ident) {
    try {
      return loadTable(ident) != null;
    } catch (NoSuchTableException e) {
      return false;
    }
  }

  /**
   * Create a table in the catalog based on the provided details.
   *
   * <p>This method returns the table as defined by the user without applying all defaults. If you
   * need the table with default values applied, use the {@link #loadTable(NameIdentifier)} method
   * after creation.
   *
   * @param ident A table identifier.
   * @param columns The columns of the new table.
   * @param comment The table comment.
   * @param properties The table properties.
   * @return The table as defined by the caller, without all default values.
   * @throws NoSuchSchemaException If the schema does not exist.
   * @throws TableAlreadyExistsException If the table already exists.
   */
  default Table createTable(
      NameIdentifier ident, Column[] columns, String comment, Map<String, String> properties)
      throws NoSuchSchemaException, TableAlreadyExistsException {
    return createTable(
        ident,
        columns,
        comment,
        properties,
        Transforms.EMPTY_TRANSFORM,
        Distributions.NONE,
        new SortOrder[0]);
  }

  /**
   * Create a table in the catalog.
   *
   * @param ident A table identifier.
   * @param columns The columns of the new table.
   * @param comment The table comment.
   * @param properties The table properties.
   * @param partitions The table partitioning.
   * @return The created table metadata.
   * @throws NoSuchSchemaException If the schema does not exist.
   * @throws TableAlreadyExistsException If the table already exists.
   */
  default Table createTable(
      NameIdentifier ident,
      Column[] columns,
      String comment,
      Map<String, String> properties,
      Transform[] partitions)
      throws NoSuchSchemaException, TableAlreadyExistsException {
    return createTable(
        ident, columns, comment, properties, partitions, Distributions.NONE, new SortOrder[0]);
  }

  /**
   * Create a table in the catalog.
   *
   * @param ident A table identifier.
   * @param columns The columns of the new table.
   * @param comment The table comment.
   * @param properties The table properties.
   * @param distribution The distribution of the table
   * @param sortOrders The sort orders of the table
   * @return The created table metadata.
   * @throws NoSuchSchemaException If the schema does not exist.
   * @throws TableAlreadyExistsException If the table already exists.
   */
  default Table createTable(
      NameIdentifier ident,
      Column[] columns,
      String comment,
      Map<String, String> properties,
      Distribution distribution,
      SortOrder[] sortOrders)
      throws NoSuchSchemaException, TableAlreadyExistsException {
    return createTable(
        ident, columns, comment, properties, new Transform[0], distribution, sortOrders);
  }

  /**
   * Create a table in the catalog.
   *
   * @param ident A table identifier.
   * @param columns The columns of the new table.
   * @param comment The table comment.
   * @param properties The table properties.
   * @param distribution The distribution of the table.
   * @return The created table metadata.
   * @throws NoSuchSchemaException If the schema does not exist.
   * @throws TableAlreadyExistsException If the table already exists.
   */
  default Table createTable(
      NameIdentifier ident,
      Column[] columns,
      String comment,
      Map<String, String> properties,
      Distribution distribution)
      throws NoSuchSchemaException, TableAlreadyExistsException {
    return createTable(
        ident, columns, comment, properties, new Transform[0], distribution, new SortOrder[0]);
  }

  /**
   * Create a table in the catalog.
   *
   * @param ident A table identifier.
   * @param columns The columns of the new table.
   * @param comment The table comment.
   * @param properties The table properties.
   * @param sortOrders The sort orders of the table
   * @return The created table metadata.
   * @throws NoSuchSchemaException If the schema does not exist.
   * @throws TableAlreadyExistsException If the table already exists.
   */
  default Table createTable(
      NameIdentifier ident,
      Column[] columns,
      String comment,
      Map<String, String> properties,
      SortOrder[] sortOrders)
      throws NoSuchSchemaException, TableAlreadyExistsException {
    return createTable(
        ident, columns, comment, properties, new Transform[0], Distributions.NONE, sortOrders);
  }

  /**
   * Create a partitioned table in the catalog.
   *
   * @param ident A table identifier.
   * @param columns The columns of the new table.
   * @param comment The table comment.
   * @param properties The table properties.
   * @param distribution The distribution of the table
   * @param sortOrders The sort orders of the table
   * @param partitions The table partitioning.
   * @return The created table metadata.
   * @throws NoSuchSchemaException If the schema does not exist.
   * @throws TableAlreadyExistsException If the table already exists.
   */
  default Table createTable(
      NameIdentifier ident,
      Column[] columns,
      String comment,
      Map<String, String> properties,
      Transform[] partitions,
      Distribution distribution,
      SortOrder[] sortOrders)
      throws NoSuchSchemaException, TableAlreadyExistsException {
    return createTable(
        ident,
        columns,
        comment,
        properties,
        partitions,
        distribution,
        sortOrders,
        Indexes.EMPTY_INDEXES);
  }

  /**
   * Create a table in the catalog.
   *
   * @param ident A table identifier.
   * @param columns The columns of the new table.
   * @param comment The table comment.
   * @param properties The table properties.
   * @param distribution The distribution of the table
   * @param sortOrders The sort orders of the table
   * @param partitions The table partitioning.
   * @param indexes The table indexes.
   * @return The created table metadata.
   * @throws NoSuchSchemaException If the schema does not exist.
   * @throws TableAlreadyExistsException If the table already exists.
   */
  Table createTable(
      NameIdentifier ident,
      Column[] columns,
      String comment,
      Map<String, String> properties,
      Transform[] partitions,
      Distribution distribution,
      SortOrder[] sortOrders,
      Index[] indexes)
      throws NoSuchSchemaException, TableAlreadyExistsException;

  /**
   * Apply the {@link TableChange change} to a table in the catalog.
   *
   * <p>Implementations may reject the change. If any change is rejected, no changes should be
   * applied to the table.
   *
   * @param ident A table identifier.
   * @param changes Table changes to apply to the table.
   * @return The updated table metadata.
   * @throws NoSuchTableException If the table does not exist.
   * @throws IllegalArgumentException If the change is rejected by the implementation.
   */
  Table alterTable(NameIdentifier ident, TableChange... changes)
      throws NoSuchTableException, IllegalArgumentException;

  /**
   * Removes both the metadata and the directory associated with the table from the file system if
   * the table is not an external table. In case of an external table, only the associated metadata
   * is removed.
   *
   * @param ident A table identifier.
   * @return True if the table is dropped, false if the table does not exist.
   */
  boolean dropTable(NameIdentifier ident);

  /**
   * Drop a table from the catalog and completely remove its data. Removes both the metadata and the
   * directory associated with the table completely and skipping trash. If the table is an external
   * table or the catalogs don't support purge table, {@link UnsupportedOperationException} is
   * thrown.
   *
   * <p>If the catalog supports to purge a table, this method should be overridden. The default
   * implementation throws an {@link UnsupportedOperationException}.
   *
   * @param ident A table identifier.
   * @return True if the table is purged, false if the table does not exist.
   * @throws UnsupportedOperationException If the catalog does not support to purge a table.
   */
  default boolean purgeTable(NameIdentifier ident) throws UnsupportedOperationException {
    throw new UnsupportedOperationException("purgeTable not supported.");
  }
}
